setup.py

```diff
@@ -1,48 +1,98 @@
 from __future__ import absolute_import, division, print_function
 
-from setuptools import setup, find_packages, Extension
+from distutils.command.sdist import sdist as _sdist
+import glob
 import os
 
-import numpy as np
-
-from Cython.Build import cythonize
+from setuptools import setup, find_packages, Extension
+from setuptools.command.build_ext import build_ext as _build_ext
 
+PACKAGE_NAME = 'stratify'
+PACKAGE_DIR = os.path.abspath(os.path.dirname(__file__))
+CYTHON_DIRECTIVES = {'linetrace': True, 'binding': True}
 
-NAME = 'stratify'
-DIR = os.path.abspath(os.path.dirname(__file__))
+try:
+    # Detect if Cython is available. Where it is, use it, otherwise fall back
+    # to C source included with source distribution.
+    import Cython
+    USE_CYTHON = True
+except ImportError:
+    USE_CYTHON = False
 
-extension_kwargs = {'include_dirs': [np.get_include()]}
+source_suffix = '.pyx' if USE_CYTHON else '.c'
+extension_kwargs = {}
 cython_coverage_enabled = os.environ.get('CYTHON_COVERAGE', None)
-if cython_coverage_enabled:
+if USE_CYTHON and cython_coverage_enabled:
     extension_kwargs.update({'define_macros': [('CYTHON_TRACE_NOGIL', '1')]})
+if USE_CYTHON:
+    # Cython requires numpy include headers
+    import numpy as np
+    extension_kwargs.update({'include_dirs': [np.get_include()]})
+
+extensions = []
+cmdclass = {}
+for source_file in glob.glob('{}/*{}'.format(PACKAGE_NAME, source_suffix)):
+    source_file_nosuf, _ = os.path.splitext(os.path.basename(source_file))
+    extensions.append(
+        Extension('{}.{}'.format(PACKAGE_NAME, source_file_nosuf),
+                  sources=['{}/{}{}'.format(
+                      PACKAGE_NAME, source_file_nosuf, source_suffix)],
+                  **extension_kwargs))
+
+if USE_CYTHON:
+    from Cython.Build import cythonize
+    extensions = cythonize(extensions, compiler_directives=CYTHON_DIRECTIVES)
+
+
+class SDist(_sdist):
+    # Source distribution build runs Cython so that Cython is not needed as
+    # an install dependency.
+    def run(self):
+        from Cython.Build import cythonize
+        cythonize(extensions, compiler_directives=CYTHON_DIRECTIVES)
+        _sdist.run(self)
+
+
+cmdclass['sdist'] = SDist
+
+
+def numpy_build_ext(pars):
+    # Delay numpy import so that setup.py can be run without numpy already
+    # being installed.
+    class NumpyBuildExt(_build_ext):
+        def finalize_options(self):
+            _build_ext.finalize_options(self)
+            __builtins__.__NUMPY_SETUP__ = False
+            import numpy
+            self.include_dirs.append(numpy.get_include())
+
+    return NumpyBuildExt(pars)
```

Comment:
@tjtg Would you mind rejigging the order of some definitions...

Typically, we have the following order:

- imports
- globals
- classes
- functions
- module code

That said, could you move the `SDist` class to just after the `try: except`, then followed by the `numpy_build_ext` and `extract_version` functions, thanks